README.md
# Gencal is a Python 3.6 Tkinter-based Gcode generator.

Gencal is a GUI based tool, used to generate gcode for a GRBL controlled
Laser engraver. This gcode is used to test to generate the power and
speed settings for a Laser engraver project. It can also be used in the
XY calabration setting of the Laser engraver.

![GenCal Main Window](https://github.com/wrw1/gencal/blob/master/Icons/mainWindow.gif)

To run it, navigate to the directory
*containing* "gencal.py", and run:

    python3 gencal.py



NEWS.txt
Gencal 1.0 Initial release on 8/20/2018


help.py
""" help.py: Implement the Gencal help menu.
Contents are subject to revision at any time, without notice.

Help => Gencal Help: Display help.html with proper formatting.

HelpParser - Parse help.html and and render to tk Text.

HelpText - Display formatted help.html.

HelpFrame - Contain text, scrollbar, and table-of-contents.
(This will be needed for display in a future tabbed window.)

HelpWindow - Display HelpFrame in a standalone window.

copy_strip - Copy idle.html to help.html, rstripping each line.

show - Create HelpWindow.  Called in MainWindow.help_dialog.
"""
from html.parser import HTMLParser
from os.path import abspath, dirname, isdir, isfile, join
from tkinter import Tk, Toplevel, Frame, Text, Scrollbar, Menu, Menubutton, PhotoImage
from tkinter import font as tkfont
#from idlelib.configHandler import idleConf

use_ttk = False # until available to import
if use_ttk:
    from tkinter.ttk import Menubutton

from globals import resource_path

## About Gencal ##


## Gencal Help ##

class HelpParser(HTMLParser):
    """Render help.html into a text widget.

    The overridden handle_xyz methods handle a subset of html tags.
    The supplied text should have the needed tag configurations.
    The behavior for unsupported tags, such as table, is undefined.
    """
    def __init__(self, text):
        HTMLParser.__init__(self, convert_charrefs=True)
        self.text = text         # text widget we're rendering into
        self.tags = ''           # current block level text tags to apply
        self.chartags = ''       # current character level text tags
        self.show = False        # used so we exclude page navigation
        self.hdrlink = False     # used so we don't show header links
        self.level = 0           # indentation level
        self.pre = False         # displaying preformatted text
        self.hprefix = ''        # prefix such as '25.5' to strip from headings
        self.nested_dl = False   # if we're in a nested <dl>
        self.simplelist = False  # simple list (no double spacing)
        self.toc = []            # pair headers with text indexes for toc
        self.header = ''         # text within header tags for toc
        self.imgFN = ''

    def indent(self, amt=1):
        self.level += amt
        self.tags = '' if self.level == 0 else 'l'+str(self.level)

    def handle_starttag(self, tag, attrs):
        "Handle starttags in help.html."
        fn = ''
        class_ = ''
        for a, v in attrs:
            if a == 'class':
                class_ = v
            if a == 'src':
                fn = v
        s = ''
        if tag == 'div' and class_ == 'section':
            self.show = True    # start of main content
        elif tag == 'div' and class_ == 'sphinxsidebar':
            self.show = False   # end of main content
        elif tag == 'p' and class_ != 'first':
            s = '\n'
        elif tag == 'span' and class_ == 'pre':
            self.chartags = 'pre'
        elif tag == 'span' and class_ == 'versionmodified':
            self.chartags = 'em'
        elif tag == 'em':
            self.chartags = 'em'
        elif tag == 'strong':
            self.chartags = 'strong'
        elif tag in ['ul', 'ol']:
            self.simplelist = True
            '''
            if class_.find('simple') != -1:
                s = '\n'
                self.simplelist = True
            else:
                self.simplelist = False
            '''
            self.indent()
        elif tag == 'dl':
            if self.level > 0:
                self.nested_dl = True
        elif tag == 'li':
            s = '\n* ' if self.simplelist else '\n\n* '
        elif tag == 'dt':
            s = '\n\n' if not self.nested_dl else '\n'  # avoid extra line
            self.nested_dl = False
        elif tag == 'dd':
            self.indent()
            s = '\n'
        elif tag == 'pre':
            self.pre = True
            if self.show:
                self.text.insert('end', '\n\n')
            self.tags = 'preblock'
        elif tag == 'a' and class_ == 'headerlink':
            self.hdrlink = True
        elif tag == 'h1':
            self.tags = tag
        elif tag in ['h2', 'h3']:
            if self.show:
                self.header = ''
                self.text.insert('end', '\n\n')
            self.tags = tag
        elif tag == 'img':
            self.text.insert('end','\n')
            pic=PhotoImage(file=fn)
            box = ("{blue blue blue blue}")
            pic.put(box, to=(20,20,20,20))
            self.text.image_create('end', image=pic)
            s = '\n'
        elif tag == 'br':
            s='\n'

        if self.show:
            self.text.insert('end', s, (self.tags, self.chartags))

    def handle_endtag(self, tag):
        "Handle endtags in help.html."
        if tag in ['h1', 'h2', 'h3']:
            self.indent(0)  # clear tag, reset indent
            if self.show:
                self.toc.append((self.header, self.text.index('insert')))
        elif tag in ['span', 'em', 'strong']:
            self.chartags = ''
        elif tag == 'a':
            self.hdrlink = False
        elif tag == 'pre':
            self.pre = False
            self.tags = ''
        elif tag in ['ul', 'dd', 'ol']:
            self.indent(amt=-1)

    def handle_data(self, data):
        "Handle date segments in help.html."
        if self.show and not self.hdrlink:
            d = data if self.pre else data.replace('\n', ' ')
            if self.tags == 'h1':
                self.hprefix = d[0:d.index(' ')]
            if self.tags in ['h1', 'h2', 'h3'] and self.hprefix != '':
                if d[0:len(self.hprefix)] == self.hprefix:
                    d = d[len(self.hprefix):].strip()
                self.header += d
            self.text.insert('end', d, (self.tags, self.chartags))


class HelpText(Text):
    "Display help.html."
    def __init__(self, parent, filename):
        "Configure tags and feed file to parser."
        # uwide = idleConf.GetOption('main', 'EditorWindow', 'width', type='int')
        # uhigh = idleConf.GetOption('main', 'EditorWindow', 'height', type='int')
        uwide = 80
        uhigh = 40
        uhigh = 3 * uhigh // 4  # lines average 4/3 of editor line height
        Text.__init__(self, parent, wrap='word', highlightthickness=0,
                      padx=5, borderwidth=0, width=uwide, height=uhigh)

        normalfont = self.findfont(['TkDefaultFont', 'arial', 'helvetica'])
        fixedfont = self.findfont(['TkFixedFont', 'monaco', 'courier'])
        self['font'] = (normalfont, 12)
        self.tag_configure('em', font=(normalfont, 12, 'italic'))
        self.tag_configure('strong', font=(normalfont, 12, 'bold'))
        self.tag_configure('h1', font=(normalfont, 20, 'bold'))
        self.tag_configure('h2', font=(normalfont, 18, 'bold'))
        self.tag_configure('h3', font=(normalfont, 15, 'bold'))
        self.tag_configure('pre', font=(fixedfont, 12), background='#f6f6ff')
        self.tag_configure('preblock', font=(fixedfont, 10), lmargin1=25,
                borderwidth=1, relief='solid', background='#eeffcc')
        self.tag_configure('l1', lmargin1=25, lmargin2=25)
        self.tag_configure('l2', lmargin1=50, lmargin2=50)
        self.tag_configure('l3', lmargin1=75, lmargin2=75)
        self.tag_configure('l4', lmargin1=100, lmargin2=100)

        self.parser = HelpParser(self)
        with open(filename, 'r') as f:
            contents = f.read()
        self.parser.feed(contents)
        self['state'] = 'disabled'

    def findfont(self, names):
        "Return name of first font family derived from names."
        for name in names:
            if name.lower() in (x.lower() for x in tkfont.names(root=self)):
                font = tkfont.Font(name=name, exists=True, root=self)
                return font.actual()['family']
            elif name.lower() in (x.lower()
                                  for x in tkfont.families(root=self)):
                return name


class HelpFrame(Frame):
    "Display html text, scrollbar, and toc."
    def __init__(self, parent, filename):
        Frame.__init__(self, parent)
        text = HelpText(self, filename)
        self['background'] = text['background']
        scroll = Scrollbar(self, command=text.yview)
        text['yscrollcommand'] = scroll.set
        self.rowconfigure(0, weight=1)
        self.columnconfigure(1, weight=1)  # text
        self.toc_menu(text).grid(column=0, row=0, sticky='nw')
        text.grid(column=1, row=0, sticky='nsew')
        scroll.grid(column=2, row=0, sticky='ns')

    def toc_menu(self, text):
        "Create table of contents as drop-down menu."
        toc = Menubutton(self, text='TOC')
        drop = Menu(toc, tearoff=False)
        for lbl, dex in text.parser.toc:
            drop.add_command(label=lbl, command=lambda dex=dex:text.yview(dex))
        toc['menu'] = drop
        return toc


class HelpWindow(Toplevel):
    "Display frame with rendered html."
    def __init__(self, parent, filename, title):
        Toplevel.__init__(self, parent)
        self.wm_title(title)
        self.protocol("WM_DELETE_WINDOW", self.destroy)
        HelpFrame(self, filename).grid(column=0, row=0, sticky='nsew')
        self.grid_columnconfigure(0, weight=1)
        self.grid_rowconfigure(0, weight=1)



def show_gencalhelp(parent):
    "Create HelpWindow; called from Idle Help event handler."
    #filename = rejoin(abspath(dirname(__file__)), 'help.html')
    filename = resource_path('help.html')
    if not isfile(filename):
        # try copy_strip, present message
        return
    HelpWindow(parent, filename, 'Gencal Help')

# if __name__ == '__main__':
#     from idlelib.idle_test.htest import run
#     run(show_idlehelp)


MainWindow.py
from tkinter import *
import tkinter.messagebox
import tkinter.filedialog
import aboutDialog
from help import *
import globals as G

# Here, we are creating our class, Window, and inheriting from the Frame
# class. Frame is a class from the tkinter module. (see Lib/tkinter/__init__)
class Window(Frame):

    # Define settings upon initialization. Here you can specify
	def __init__(self, master=None):
		# parameters that you want to send through the Frame class.
		Frame.__init__(self, master)
		self.master = master
		self.init_window()
		self.about = aboutDialog
		self.gencalhelp = HelpWindow

	#Creation of init_window
	def init_window(self):

		# changing the title of our master widget
		self.master.title("Gencal " + G.Version)

		# allowing the widget to take the full space of the root window
		self.pack(fill=BOTH, expand=1)

		# creating a menu instance
		menu = Menu(self.master)
		self.master.config(menu=menu)

		# create the file pulldown menu
		filemenu = Menu(menu, tearoff=0)
		filemenu.add_command(label="Save As ...", underline=0, accelerator="Ctrl-s",
							 command=self.save_file)
		filemenu.add_separator()
		filemenu.add_command(label="Exit", command=self.client_exit)
		menu.add_cascade(label="File", menu=filemenu)

		# create the help pulldown menu
		helpmenu = Menu(menu, tearoff=0)
		helpmenu.add_command(label="About Gencal", command=self.about_dialog)
		helpmenu.add_separator()
		helpmenu.add_command(label="Gencal Help", command=self.help_dialog)
		menu.add_cascade(label="Help", menu=helpmenu)

	def about_dialog(self):
		aboutDialog.show(self)

	def help_dialog(self):
		show_gencalhelp(self)

	def about(self):
		lines = 'Gencal\n' + '(Generate Calabration Gcode)\n' + 'Version 1.0'
		tkinter.messagebox.showinfo('About', lines)

	def save_file(self):
		global gcode
		options =  {}
		options['defaultextension'] = '.nc'
		options['filetypes'] = [('all files', '.*'), ('gcode files', '.nc'), ('gcode files', '.gcode')]
		options['initialfile'] = 'test.nc'
		options['title'] = 'Save Gcode as...'

		fn = tkinter.filedialog.asksaveasfilename(**options)
		if not fn:
			return
		try:
			UseFile = open(fn, 'w')
			for line in G.gcode:
				UseFile.write(line+'\n')
			UseFile.close()
		except:
			tkinter.messagebox.showinfo("Error", "Unable to open to write " + fn)

	def client_exit(self):
		exit()


ScrollableFrame.py
from tkinter import *

# http://tkinter.unpythonic.net/wiki/VerticalScrolledFrame

class VerticalScrolledFrame(Frame):
	"""A pure Tkinter scrollable frame that actually works!
	* Use the 'interior' attribute to place widgets inside the scrollable frame
	* Construct and pack/place/grid normally
	* This frame only allows vertical scrolling

	"""
	def __init__(self, parent, *args, **kw):
		Frame.__init__(self, parent, *args, **kw)

		# create a canvas object and a vertical scrollbar for scrolling it
		vscrollbar = Scrollbar(self, orient=VERTICAL)
		vscrollbar.pack(fill=Y, side=RIGHT, expand=FALSE)
		canvas = Canvas(self, bd=0, highlightthickness=0,
						yscrollcommand=vscrollbar.set)
		canvas.pack(side=LEFT, fill=BOTH, expand=TRUE)
		vscrollbar.config(command=canvas.yview)

		# reset the view
		canvas.xview_moveto(0)
		canvas.yview_moveto(0)

		# create a frame inside the canvas which will be scrolled with it
		self.interior = interior = Frame(canvas)
		interior_id = canvas.create_window(0, 0, window=interior,
										   anchor=NW)

		# track changes to the canvas and frame width and sync them,
		# also updating the scrollbar
		def _configure_interior(event):
			# update the scrollbars to match the size of the inner frame
			size = (interior.winfo_reqwidth(), interior.winfo_reqheight())
			canvas.config(scrollregion="0 0 %s %s" % size)
			if interior.winfo_reqwidth() != canvas.winfo_width():
				# update the canvas's width to fit the inner frame
				canvas.config(width=interior.winfo_reqwidth())
		interior.bind('<Configure>', _configure_interior)

		def _configure_canvas(event):
			if interior.winfo_reqwidth() != canvas.winfo_width():
				# update the inner frame's width to fill the canvas
				canvas.itemconfigure(interior_id, width=canvas.winfo_width())
		canvas.bind('<Configure>', _configure_canvas)




aboutDialog.py
"""About Dialog for gencal

"""

import os
from sys import version
from tkinter import *
import globals as G
from globals import resource_path

#from idlelib import ui


dlg = None

def show(parent):
    "Main routine; show dialog window, creating or raising as necessary."
    global dlg
    if dlg is None:
        dlg = AboutDialog(parent, destroy_callback=_destroyed)
    dlg.lift()
    dlg.focus_set()

def _destroyed():
    global dlg
    dlg = None


class AboutDialog(Toplevel):
    """About dialog for IDLE """
    def __init__(self, parent, title='About Gencal', _htest=False,
                 destroy_callback=None):
        """
        _htest - bool, change box location when running htest
        """
        Toplevel.__init__(self, parent)
        self.destroy_callback = destroy_callback
        self.configure(borderwidth=5)
        # place dialog below parent if running htest
        self.geometry("+%d+%d" % (
                        parent.winfo_rootx()+30,
                        parent.winfo_rooty()+(30 if not _htest else 100)))
        self.bg = "#bbbbbb"
        self.fg = "#000000"
#        self.link_cursor = ui.clickable_cursor
        self.CreateWidgets()
        self.resizable(height=FALSE, width=FALSE)
        self.title(title)
        self.protocol("WM_DELETE_WINDOW", self.close)
        self.parent = parent
        self.bind('<Escape>', self.close)   # dismiss dialog

    def CreateWidgets(self):
        self['borderwidth'] = 0
        release = version[:version.index(' ')]
        '''
        logofn = os.path.join(os.path.abspath(os.path.dirname(__file__)),
                              "Icons", "gencal_48.gif")
        '''
        logofn = resource_path('Icons/gencal_48.gif')
        self.picture = PhotoImage(master=self._root(), file=logofn)
        self.frameBg = frameBg = Frame(self, bg=self.bg, borderwidth=0)
        frameBg.grid(sticky='nsew')
        labelTitle = Label(frameBg, text='Gencal ' + G.Version, fg=self.fg, bg=self.bg,
                           font=('courier', 24, 'bold'))
        labelTitle.grid(row=0, column=1, sticky=W, padx=10, pady=[10,0])
        labelPicture = Label(frameBg, image=self.picture, bg=self.bg)
        labelPicture.grid(row=0, column=0, sticky=NE, rowspan=2,
                          padx=10, pady=10)
        byline = "Generate Calabration Gcode"
        labelDesc = Label(frameBg, text=byline, justify=LEFT,
                          fg=self.fg, bg=self.bg)
        labelDesc.grid(row=1, column=1, sticky=W, columnspan=3, padx=10,
                       pady=[0,20])
        # labelEmail = Label(frameBg, text='email:  idle-dev@python.org',
        #                    justify=LEFT, fg=self.fg, bg=self.bg)
        # labelEmail.grid(row=6, column=1, columnspan=2,
        #                 sticky=W, padx=10, pady=0)
        # labelWWW = Label(frameBg, text='https://docs.python.org/' +
        #                  version[:3] + '/library/idle.html',
        #                  justify=LEFT, fg=self.fg, bg=self.bg)
        #labelWWW.grid(row=7, column=1, columnspan=2, sticky=W, padx=10, pady=0)
        tkVer = self.tk.call('info', 'patchlevel')
        labelVersion = Label(frameBg, text='Python ' +
                             release + '     (with Tk '+tkVer+')',
                             fg=self.fg, bg=self.bg)
        labelVersion.grid(row=4, column=1, sticky=W, padx=10, pady=[0,5])
        self.morelink = Label(frameBg, text='More...', fg='blue', bg=self.bg,
                              cursor='hand2')
        self.morelink.grid(column=0, columnspan=3, pady=10, padx=10, sticky=E)
        self.morelink.bind('<1>', self.showMore)

    def showMore(self, ev=None):
        self.morelink.grid_forget()
        fmore = Frame(self.frameBg, borderwidth=2, relief='ridge', bg='white')
        fmore.grid_columnconfigure(0, weight=1)
        self.t = Text(fmore, height=15, width=80, borderwidth=0,
                      highlightthickness=0, state='disabled', bg='white')
        s = Scrollbar(fmore, command=self.t.yview)
        self.t['yscrollcommand'] = s.set
        self.load_moreinfo()
        self.which = StringVar(self)
        self.which.set(self.info[0][0])
        self.which.trace_variable('w', self.changeInfo)
        fmore.grid(column=0, row=20, sticky='nwes', padx=5, pady=[15,5],
                   columnspan=4)
        l = []
        for k, t in self.info:
            l.append(k)
        wh = OptionMenu(fmore, self.which, *l)
        wh.grid(column=0, row=0, pady=[2,10])
        self.t.grid(column=0, row=1)
        s.grid(column=1, row=0, sticky='ns', rowspan=2)
        self.changeInfo()

    def changeInfo(self, *params):
        for key, txt in self.info:
            if key == self.which.get():
                self.t['state'] = 'normal'
                self.t.delete('1.0', 'end')
                self.t.insert('1.0', txt)
                self.t['state'] = 'disabled'

    def load_moreinfo(self):
        self.info = []
        self.load_from_file('Gencal Readme', 'README.txt')
        self.load_from_file('Gencal News', 'NEWS.txt')
        self.load_from_file('Gencal ToDo', 'TODO.txt')
        self.load_from_file('Gencal Credits', 'CREDITS.txt')

    def load_from_file(self, key, filename, encoding=None):
        #fn = os.path.join(os.path.abspath(os.path.dirname(__file__)), filename)
        fn = resource_path(filename)
        try:
            with open(fn, 'r', encoding=encoding) as file:
                contents = file.read()
        except IOError:
            pass
        else:
            self.info.append((key, contents))

    def load_from_printer(self, key, printercmd):
        printercmd._Printer__setup()
        self.info.append((key, '\n'.join(printercmd._Printer__lines)))

    def close(self, event=None):
        if self.destroy_callback:
            self.destroy_callback()
        self.destroy()

# if __name__ == '__main__':
#     from idlelib.idle_test.htest import run
#     run(AboutDialog)
